### PR TITLE
Add EA.Razor APIs for breakpoint & proximity expression resolution.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Had to add add a project reference to the corresponding editor APIs in Microsoft.CodeAnalysis.CSharp.Features.csproj.
- Prefixed our ExternalAccess API's with "Razor" to not conflict in symbol recognition when working in the Roslyn solution.
- Added a non-test method to the `CSharpProximityExpressionService` (which is already internal) to expose a more first class way in getting proximity expressiosn from a syntax tree (doesn't need to be async). This API still isn't accessible on the `IProximityExpressionService` since Razor doesn't want to actually be in the business of resolving full on services and in this case doesn't actually need to.

Fixes dotnet/aspnetcore#29547

### CODE REVIEW

File: src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs [hunk 1]
```diff
@@ -100,6 +100,8 @@ public CSharpProximityExpressionsService()
             }
         }
 
+        public static IList<string> GetProximityExpressions(SyntaxTree syntaxTree, int position) => Do(syntaxTree, position);
```
Comment (NTaylorMullen) [inline src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs:103]: I could have called the "Do" API but it just didn't feel right. Up for suggestions though

File: src/Features/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.Features.csproj [hunk 2]
```diff
@@ -26,6 +26,9 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.CSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.EditorFeatures.Wpf" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.Remote.Workspaces" />
+    <!-- BEGIN External Access -->
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.Razor" />
```
Comment (NTaylorMullen) [inline src/Features/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.Features.csproj:30]: Looks like we're the first ExternalAccess member to need this. I doing something wrong? 😄

**Review by MMLaws7 — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell .  I don't love this.  I would have preferred if we had tried to EA IProximityExpressionsService instead. But if Sam is ok with this, that's fine with me.

**Comment by NTaylorMullen** [comment]
> Tagging @sharwell . I don't love this. I would have preferred if we had tried to EA IProximityExpressionsService instead. But if Sam is ok with this, that's fine with me.

I'd be willing; however that service takes in a TextDocument, something we don't have in Razor since we're spinning up these C# SyntaxTree's independent of the actual workspace. If there's a way around that restriction I'm all ears of course though 😄

**Comment by NTaylorMullen** [comment]
Ping. Also, what's the best way to investigate why the CI is failing consistently here?

File: src/Features/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.Features.csproj (hunk 2 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.Features.csproj:30]: Some external access consumers have a dedicated assembly (F#, Razor). Others use Restricted IVTs.

File: src/Tools/ExternalAccess/Razor/RazorCSharpBreakpointResolver.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+#nullable enable
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.EditAndContinue;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class RazorCSharpBreakpointResolver
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorCSharpBreakpointResolver.cs:13]:
```suggestion
    internal static class RazorBreakpointSpans
```

File: src/Tools/ExternalAccess/Razor/RazorCSharpBreakpointResolver.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public bool TryGetBreakpointSpan(SyntaxTree tree, int position, CancellationToken cancellationToken, out TextSpan breakpointSpan)
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorCSharpBreakpointResolver.cs:15]:
```suggestion
        public static bool TryGetBreakpointSpan(SyntaxTree tree, int position, CancellationToken cancellationToken, out TextSpan breakpointSpan)
```

File: src/Tools/ExternalAccess/Razor/RazorCSharpProximityExpressionResolver.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp.Debugging;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal class RazorCSharpProximityExpressionResolver
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorCSharpProximityExpressionResolver.cs:12]:
```suggestion
    internal static class RazorCSharpProximityExpressionResolverService
```

File: src/Tools/ExternalAccess/Razor/RazorCSharpProximityExpressionResolver.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public IList<string> GetProximityExpressions(SyntaxTree syntaxTree, int absoluteIndex)
+            => CSharpProximityExpressionsService.Do(syntaxTree, absoluteIndex);
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/RazorCSharpProximityExpressionResolver.cs:15]: ❗ `Do` is a private helper method. I'm going to send a separate pull request to change the accessibility to `private` and expose it to testing through a test accessor.

Comment (NTaylorMullen) [reply src/Tools/ExternalAccess/Razor/RazorCSharpProximityExpressionResolver.cs:15]: WOuld you like me to hold off on this PR until then or do it in parallel?

**Comment by NTaylorMullen** [comment]
This good to merge?

File: src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs [hunk 7]
```diff
@@ -100,6 +100,8 @@ public CSharpProximityExpressionsService()
             }
         }
 
+        public static IList<string> GetProximityExpressions(SyntaxTree syntaxTree, int position) => Do(syntaxTree, position, cancellationToken: default);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs:103]: ❔ Why is this not using one of the existing methods?

Comment (sharwell) [inline src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs:103]: If we do keep this (might be removed from https://github.com/dotnet/roslyn/pull/50718/files#r569757295), it needs to have a proper `CancellationToken` parameter (caller may pass `CancellationToken.None`, but it's not going to do so by default).

Comment (NTaylorMullen) [reply src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs:103]: So I reutilized the existing "Do" because that was the written way to do "synchronous" expression resolution. Because of the synchronous requirement I felt using it directly was the best approach. Alternatively I could have written:
```
new Worker(syntaxTree, position).Do(cancellationToken)
```

Comment (NTaylorMullen) [reply src/Features/CSharp/Portable/Debugging/CSharpProximityExpressionsService.cs:103]: Can add.

Out of curiosity, is there history behind passing in the cancellation tokens to these synchronous methods? Aka do we really expect the code to have a high perf cost? I imagine that the underlying bits aren't doing Task.Result which would then infer that they're doing "throwIfCancellationRequested" in tight loops.

**Comment by NTaylorMullen** [comment]
@sharwell ping

**Review by sharwell — APPROVED** [review]
